Distributed Autotuning

aorenste opened the pull request:
This is the initial prototype of distributed autotuning. It's intended to be a basis for iteration rather than the final end product.

Currently when we run a SPMD program we compile the ranks independently. As a result the autotuning is repeated on every rank. So for a 8-GPU program with 8 matmul operators we'll autotune 64 (8*8) times.

Distributed autotuning uses collectives to distribute the autotuning across the ranks so each rank autotunes 1/worldsize the total operators. So in our 8-GPU example we would only perform 8 autotunes total (one on each rank) rather than 64.

There are several advantages:
1. Faster autotuning times - each CPU/GPU does less work total
2. Better determinism - currently it's possible for two ranks to choose different algorithms for the same operator. With distributed autotuning we choose the algorithm once for the entire program.

Results:

In testing using llama3 8B on torchtitan max-autotune time was reduced from 52s -> 26s and exhaustive-autotuning was reduced from 2009s -> 613s.

Usage:

The feature is controlled by the environment variable TORCHINDUCTOR_DISTRIBUTED_AUTOTUNE.

Co-authored-by: @PaulZhang12 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163369



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

aorenste committed:
distributed autotuning

[ghstack-poisoned]

coconutruben commented on torch/_inductor/distributed_autotune.py:
```diff
@@ -0,0 +1,316 @@
+        raise NotImplementedError
+
+    @staticmethod
+    def serialize_choice(index: int, choice: ChoiceCaller) -> _DistributedChoice:
```
you can take a look at KernelTemplateChoice (KTC) and how that's implemented. I believe in the latest version I made the ChoiceCaller keep a reference to the KTC. If you serialize the KTC, it has everything in place to just call ktc.choice and you get the choicecaller. Everything in the ktc should be serializable, but I haven't tested that out. But the choice knows how to recreate itself from those components

one caveat is that the ktc has a reference to the template itself, which might be harder to realize if at all? but then we can just add a small layer of indirection to keep track of template uid<>templates and retrieve it again

aorenste replied:
It seems that KernelTemplateChoice isn't serializable. Pickling the template complains that "Can't pickle <function mm_grid at 0x7f42dfb4e320>: it's not the same object as torch._inductor.kernel.mm_common.mm_grid" and pickling the inputs complains that "TypeError: cannot pickle 'PyCapsule' object".

aorenste replied:
@coconutruben Take a look at the latest version - I didn't pickle the KernelTemplateChoice itself (because it can't) but I did use it on construction - is that better?

bobrenjc93 commented on test/distributed/test_dynamo_distributed.py:
```diff
@@ -1176,6 +1180,42 @@ def f(x):
 
+    @unittest.skipIf(not HAS_GPU, "Inductor+gpu needs triton and recent GPU arch")
+    @patch.object(torch._dynamo.config, "enable_compiler_collectives", True)
+    @patch.object(torch._inductor.config, "max_autotune_gemm", True)
```
as discussed you can use 

@patch.object(torch._inductor.config, "max_autotune_gemm_backends", "TRITON")

to ensure we always pick a triton template for this test

bobrenjc93 commented on torch/_inductor/distributed_autotune.py:
```diff
@@ -0,0 +1,316 @@
+            index = node.node.origin_node.meta[_DISTRIBUTED_AUTOTUNE_INDEX]
+            replacement_buf = node.node.autotune(choices_by_index[index])
+
+            out_storage = replacement_buf.data
+            assert isinstance(out_storage, StorageBox)
+            out_buffer = out_storage.data
+            assert isinstance(out_buffer, OperationBuffer)
+            assert node.node.layout == out_buffer.layout
+            replace_operation_buffer(node.node, out_buffer)
+            new_scheduler_node = scheduler.create_scheduler_node(out_buffer)
+
+            scheduler.nodes[i] = new_scheduler_node
+            scheduler.name_to_node[node.get_name()] = new_scheduler_node
+            scheduler.name_to_fused_node[node.get_name()] = new_scheduler_node
+
+            # We need to reflect the mutation renames that were recorded in the original node
+            mutation_renames = {}
+            for dep in itertools.chain(node.read_writes.reads, node.unmet_dependencies):
+                if real_name := scheduler.mutation_real_name.get(dep.name, None):
+                    mutation_renames[real_name] = dep.name
+
+            def rename_deps(deps: OrderedSet[Dep]) -> OrderedSet[Dep]:
+                return OrderedSet(dep.rename(mutation_renames) for dep in deps)
+
+            new_scheduler_node.unmet_dependencies = rename_deps(
+                new_scheduler_node.unmet_dependencies
+            )
+            new_scheduler_node.read_writes.reads = rename_deps(
+                new_scheduler_node.read_writes.reads
+            )
+
+            for new_out, old_out in zip(
+                new_scheduler_node.get_outputs(), node.get_outputs()
+            ):
+                scheduler.name_to_buf[old_out.get_name()] = new_out
+                new_out.users = old_out.users
+
+            new_scheduler_node.min_order = node.min_order
+            new_scheduler_node.max_order = node.max_order
+            new_scheduler_node.last_usage = node.last_usage
```
i see this type of logic all the time in inductor code and it seems quite fragile. ooc how did you know to do these operations (eg. rename deps, update min/max order, handle mutation reflection, etc. etc.) and how do you ensure it's exhaustive?

aorenste replied:
Lol - @PaulZhang12 wrote this part...

bobrenjc93 commented:
Do we have any initial timings? It'd be good to sanity check that this is faster. Bonus points if you have a few charts showing how the speedups change as the model increases in size

ezyang commented:
Need more PR description

aorenste committed:
Update on "distributed autotuning"

**Posted for internal discussion - not ready for general review yet.**




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

aorenste committed:
Update on "distributed autotuning"

**Posted for internal discussion - not ready for general review yet.**

Co-authored-by: Paul Zhang <paulzhanumich.edu>




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

aorenste committed:
Update on "Distributed Autotuning"

This is the initial prototype of distributed autotuning. It's intended to be a basis for iteration rather than the final end product.

Currently when we run a SPMD program we compile the ranks independently. As a result the autotuning is repeated on every rank. So for a 8-GPU program with 8 matmul operators we'll autotune 64 (8*8) times.

Distributed autotuning uses collectives to distribute the autotuning across the ranks so each rank autotunes 1/worldsize the total operators. So in our 8-GPU example we would only perform 8 autotunes total (one on each rank) rather than 64.

There are several advantages:
1. Faster autotuning times - each CPU/GPU does less work total
2. Better determinism - currently it's possible for two ranks to choose different algorithms for the same operator. With distributed autotuning we choose the algorithm once for the entire program.

Usage:

The feature is controlled by the environment variable TORCHINDUCTOR_DISTRIBUTED_AUTOTUNE.

Co-authored-by: Paul Zhang <paulzhanumich.edu>




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

PaulZhang12 commented on torch/_inductor/distributed_autotune.py:
```diff
@@ -0,0 +1,383 @@
+    ) -> dict[ChoiceCaller, float]:
+        # This should never get called. It means that a remote autotune was
+        # scheduled but never filled in.
+        raise NotImplementedError
```
nit: Maybe instead of `NotImplementedError`, throw a more informative error message?

PaulZhang12 commented on torch/_inductor/scheduler.py:
```diff
@@ -2376,6 +2380,12 @@ def _init(self, nodes: list[ir.Operation]) -> None:
+        if config.distributed_autotune:
+            from . import distributed_autotune
+
+            distributed_autotune.schedule(self)
```
Nice! This looks super clean

PaulZhang12 reviewed (approved):
LGTM is a v1! cc @eellison @coconutruben for feedback

aorenste commented on torch/_inductor/config.py:
```diff
@@ -445,6 +445,12 @@ def prologue_fusion_enabled() -> bool:
+# autotuning across distributed ranks in the same program group - so instead of
+# each rank autotuning every kernel they only autotune 1/world size kernels and
+# then share the results.
+distributed_autotune = os.environ.get("TORCHINDUCTOR_DISTRIBUTED_AUTOTUNE") == "1"
```
I'd like to eventually be able to use distributed autotuning on any kernels, not just gemms - but I guess like "max_autotune" it makes sense to have a "_gemm" specific version - and why not start with only that config so if we add non-gemms it doesn't accidentally kick it on for others...

aorenste committed:
Update on "Distributed Autotuning"

This is the initial prototype of distributed autotuning. It's intended to be a basis for iteration rather than the final end product.

Currently when we run a SPMD program we compile the ranks independently. As a result the autotuning is repeated on every rank. So for a 8-GPU program with 8 matmul operators we'll autotune 64 (8*8) times.

Distributed autotuning uses collectives to distribute the autotuning across the ranks so each rank autotunes 1/worldsize the total operators. So in our 8-GPU example we would only perform 8 autotunes total (one on each rank) rather than 64.

There are several advantages:
1. Faster autotuning times - each CPU/GPU does less work total
2. Better determinism - currently it's possible for two ranks to choose different algorithms for the same operator. With distributed autotuning we choose the algorithm once for the entire program.

Usage:

The feature is controlled by the environment variable TORCHINDUCTOR_DISTRIBUTED_AUTOTUNE.

Co-authored-by: Paul Zhang <paulzhanumich.edu>




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

aorenste committed:
Update on "Distributed Autotuning"

This is the initial prototype of distributed autotuning. It's intended to be a basis for iteration rather than the final end product.

Currently when we run a SPMD program we compile the ranks independently. As a result the autotuning is repeated on every rank. So for a 8-GPU program with 8 matmul operators we'll autotune 64 (8*8) times.

Distributed autotuning uses collectives to distribute the autotuning across the ranks so each rank autotunes 1/worldsize the total operators. So in our 8-GPU example we would only perform 8 autotunes total (one on each rank) rather than 64.

There are several advantages:
1. Faster autotuning times - each CPU/GPU does less work total
2. Better determinism - currently it's possible for two ranks to choose different algorithms for the same operator. With distributed autotuning we choose the algorithm once for the entire program.

Results:

In testing using llama3 8B on torchtitan max-autotune time was reduced from 52s -> 26s and exhaustive-autotuning was reduced from 2009s -> 613s.

Usage:

The feature is controlled by the environment variable TORCHINDUCTOR_DISTRIBUTED_AUTOTUNE.

Co-authored-by: Paul Zhang <paulzhanumich.edu>




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

aorenste commented on test/distributed/test_dynamo_distributed.py:
```diff
@@ -1200,6 +1209,124 @@ def f(x):
+    @patch.object(torch._inductor.config, "max_autotune_gemm", True)
+    @patch.object(torch._inductor.config, "distributed_max_autotune_gemm", True)
+    @parametrize("backend", ("TRITON", "ATEN"))
+    def test_multiproc_autotune_dynamic_shapes(self, backend):
```
@PaulZhang12 New dynamic shapes test

PaulZhang12 replied:
Nice! Looks good to me

aorenste committed:
Update on "Distributed Autotuning"

This is the initial prototype of distributed autotuning. It's intended to be a basis for iteration rather than the final end product.

Currently when we run a SPMD program we compile the ranks independently. As a result the autotuning is repeated on every rank. So for a 8-GPU program with 8 matmul operators we'll autotune 64 (8*8) times.

Distributed autotuning uses collectives to distribute the autotuning across the ranks so each rank autotunes 1/worldsize the total operators. So in our 8-GPU example we would only perform 8 autotunes total (one on each rank) rather than 64.

There are several advantages:
1. Faster autotuning times - each CPU/GPU does less work total
2. Better determinism - currently it's possible for two ranks to choose different algorithms for the same operator. With distributed autotuning we choose the algorithm once for the entire program.

Results:

In testing using llama3 8B on torchtitan max-autotune time was reduced from 52s -> 26s and exhaustive-autotuning was reduced from 2009s -> 613s.

Usage:

The feature is controlled by the environment variable TORCHINDUCTOR_DISTRIBUTED_AUTOTUNE.

Co-authored-by: Paul Zhang <paulzhanumich.edu>




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

aorenste committed:
Update on "Distributed Autotuning"

This is the initial prototype of distributed autotuning. It's intended to be a basis for iteration rather than the final end product.

Currently when we run a SPMD program we compile the ranks independently. As a result the autotuning is repeated on every rank. So for a 8-GPU program with 8 matmul operators we'll autotune 64 (8*8) times.

Distributed autotuning uses collectives to distribute the autotuning across the ranks so each rank autotunes 1/worldsize the total operators. So in our 8-GPU example we would only perform 8 autotunes total (one on each rank) rather than 64.

There are several advantages:
1. Faster autotuning times - each CPU/GPU does less work total
2. Better determinism - currently it's possible for two ranks to choose different algorithms for the same operator. With distributed autotuning we choose the algorithm once for the entire program.

Results:

In testing using llama3 8B on torchtitan max-autotune time was reduced from 52s -> 26s and exhaustive-autotuning was reduced from 2009s -> 613s.

Usage:

The feature is controlled by the environment variable TORCHINDUCTOR_DISTRIBUTED_AUTOTUNE.

Co-authored-by: PaulZhang12 




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

bdhirsh commented:
yep, we (still) don't have an official "SPMD mode" for the compiler. At best, though I imagine some kind of SPMD mode consisting of:

(1) the user providing a way to convince us that their compiled region is SPMD (via flag or something)

(2) possibly a debug mode where we compare the graphs across ranks at compile time and yell if they were wrong

(3) a bunch of configs we can recommend that they can turn on for additional optimization, if their region of code is SPMD.

(4) ensure that all of our default compiler passes are "pessimistic" and resilient to not blowing up when encountering non-SPMD graphs.

We made some progress on (4) earlier this year by fixing a few examples of hangs in the compiler. We haven't really don't (1) or (2) at all, since you can mostly get away with just reaping the benefits of turning on the configs from (3) if you know better. At least in terms of near term things that I'd like to do, I want to finish properly landing  https://github.com/pytorch/pytorch/pull/150302 (which is an example of an optimization you can run if you know you are SPMD)

aorenste committed:
Update on "Distributed Autotuning"

This is the initial prototype of distributed autotuning. It's intended to be a basis for iteration rather than the final end product.

Currently when we run a SPMD program we compile the ranks independently. As a result the autotuning is repeated on every rank. So for a 8-GPU program with 8 matmul operators we'll autotune 64 (8*8) times.

Distributed autotuning uses collectives to distribute the autotuning across the ranks so each rank autotunes 1/worldsize the total operators. So in our 8-GPU example we would only perform 8 autotunes total (one on each rank) rather than 64.

There are several advantages:
1. Faster autotuning times - each CPU/GPU does less work total
2. Better determinism - currently it's possible for two ranks to choose different algorithms for the same operator. With distributed autotuning we choose the algorithm once for the entire program.

Results:

In testing using llama3 8B on torchtitan max-autotune time was reduced from 52s -> 26s and exhaustive-autotuning was reduced from 2009s -> 613s.

Usage:

The feature is controlled by the environment variable TORCHINDUCTOR_DISTRIBUTED_AUTOTUNE.

Co-authored-by: PaulZhang12 




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]

aorenste committed:
Update on "Distributed Autotuning"

This is the initial prototype of distributed autotuning. It's intended to be a basis for iteration rather than the final end product.

Currently when we run a SPMD program we compile the ranks independently. As a result the autotuning is repeated on every rank. So for a 8-GPU program with 8 matmul operators we'll autotune 64 (8*8) times.

Distributed autotuning uses collectives to distribute the autotuning across the ranks so each rank autotunes 1/worldsize the total operators. So in our 8-GPU example we would only perform 8 autotunes total (one on each rank) rather than 64.

There are several advantages:
1. Faster autotuning times - each CPU/GPU does less work total
2. Better determinism - currently it's possible for two ranks to choose different algorithms for the same operator. With distributed autotuning we choose the algorithm once for the entire program.

Results:

In testing using llama3 8B on torchtitan max-autotune time was reduced from 52s -> 26s and exhaustive-autotuning was reduced from 2009s -> 613s.

Usage:

The feature is controlled by the environment variable TORCHINDUCTOR_DISTRIBUTED_AUTOTUNE.

Co-authored-by: PaulZhang12 




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]
